cmd/query/app/apiv3/otlp_translator.go

```diff
@@ -0,0 +1,364 @@
+			}
+		case model.ValueType_FLOAT64:
+			kv.Value.Value = &commonv1.AnyValue_DoubleValue{
+				DoubleValue: tag.GetVFloat64(),
+			}
+		case model.ValueType_BINARY:
+			kv.Value.Value = &commonv1.AnyValue_BytesValue{
+				BytesValue: tag.GetVBinary(),
+			}
+		default:
+			kv.Value.Value = &commonv1.AnyValue_StringValue{
+				StringValue: tag.String(),
+			}
+		}
+		kvs = append(kvs, kv)
+	}
+	return kvs
+}
+
+func jLogsToOTLP(logs []model.Log) []*v1.Span_Event {
+	events := make([]*v1.Span_Event, len(logs))
+	for i, l := range logs {
+
+		var name string
+		var ignoreKeys map[string]bool
+		if messageTag, ok := model.KeyValues(l.GetFields()).FindByKey(tracetranslator.TagMessage); ok {
+			name = messageTag.GetVStr()
+			ignoreKeys = map[string]bool{}
+			ignoreKeys[tracetranslator.TagMessage] = true
+		}
+
+		events[i] = &v1.Span_Event{
+			TimeUnixNano: uint64(l.GetTimestamp().UnixNano()),
+			Name:         name,
+			Attributes:   jTagsToOTLP(l.GetFields(), ignoreKeys),
+		}
+	}
+	return events
+}
+
+func jReferencesToOTLP(refs []model.SpanRef, excludeParentID model.SpanID) []*v1.Span_Link {
+	if len(refs) == 0 || len(refs) == 1 && refs[0].SpanID == excludeParentID && refs[0].RefType == model.ChildOf {
+		return nil
+	}
+	var links []*v1.Span_Link
+	for _, r := range refs {
+		if r.SpanID == excludeParentID && r.GetRefType() == model.ChildOf {
+			continue
+		}
+		links = append(links, &v1.Span_Link{
+			TraceId: uint64ToTraceID(r.TraceID.High, r.TraceID.Low),
+			SpanId:  uint64ToSpanID(uint64(r.SpanID)),
+		})
+	}
+
+	return links
+}
+
+func getSpanStatus(tags []model.KeyValue) (*v1.Status, map[string]bool) {
+	statusCode := v1.Status_STATUS_CODE_UNSET
+	statusMessage := ""
+	statusExists := false
+
+	ignoreKeys := map[string]bool{}
+	kvs := model.KeyValues(tags)
+	if _, ok := kvs.FindByKey(tracetranslator.TagError); ok {
+		statusCode = v1.Status_STATUS_CODE_ERROR
+		statusExists = true
+		ignoreKeys[tracetranslator.TagError] = true
+	}
+	if tag, ok := kvs.FindByKey(tracetranslator.TagStatusCode); ok {
+		statusExists = true
+		if code, err := getStatusCodeValFromTag(tag); err == nil {
+			statusCode = v1.Status_StatusCode(code)
+			ignoreKeys[tracetranslator.TagStatusCode] = true
+		}
+		if tag, ok := kvs.FindByKey(tracetranslator.TagStatusMsg); ok {
+			statusMessage = tag.GetVStr()
+			ignoreKeys[tracetranslator.TagStatusMsg] = true
+		}
+	} else if tag, ok := kvs.FindByKey(conventions.AttributeHTTPStatusCode); ok {
+		statusExists = true
+		if code, err := getStatusCodeFromHTTPStatusTag(tag); err == nil {
+			// Do not set status code in case it was set to Unset.
+			if v1.Status_StatusCode(code) != v1.Status_STATUS_CODE_UNSET {
+				statusCode = v1.Status_StatusCode(code)
+			}
+
+			if tag, ok := kvs.FindByKey(tracetranslator.TagHTTPStatusMsg); ok {
+				statusMessage = tag.GetVStr()
+			}
+		}
+	}
+
+	if statusExists {
+		return &v1.Status{
+			Code:    statusCode,
+			Message: statusMessage,
+		}, ignoreKeys
+	}
+	return nil, nil
```

Comment:
why not `return nil, ignoreKeys`? That way, you no longer need [the `nil` check block](https://github.com/jaegertracing/jaeger/pull/3086/files#diff-91bb46c6591aeca3f2c806023af52ba7df5ef679e8e808dcc0f0ae58d1a6eeb0R84-R86).